File: worker/activity.ts
import { Observable } from "rxjs";

export interface Metadata {
  type: string;
  name: string;
}


// action 的返回值。
// 有些动作没有返回，比如sleep
// 业务act Result必须要能反映是否失败，如果失败，必须有失败信息；如果成功，需要有数据?。
export class ActionResult {
  // // 动作基本信息
  // // metadata: Metadata;
  // // 错误信息。null表示没有错误
  // error: any;

  // constructor(public metadata: Metadata, err?: any) {
  //   // this.metadata = metadata;

  //   if (!err || err.ok) {
  //     this.error = null;
  //   } else {
  //     this.error = err;
  //   }
  // }

  // 一个动作的结果包含3个信息：
  // 1. 是哪个动作
  // 2. 调用的参数是啥
  // 3. 结果是啥
  constructor(public result: any, public action: any, public args: any) {}

  ok(): boolean {
    return !this.result.error;
  }

}

export interface Activity {
  parse(data: any): void;

  // 每个act执行结果不一定是一个结果，因为有类似loop这种act存在。
  proceed(ctx: any): Observable<ActionResult>;
}

File: worker_java/src/main/kotlin/com/tac/whitebox/Activity.kt
package com.tac.whitebox

import io.reactivex.Maybe
import io.reactivex.Observable
import org.apache.commons.logging.LogFactory
import org.springframework.stereotype.Component

interface ActivityResult

interface ActivityContext

interface Activity {
    fun proceed(ctx: ActivityContext): Observable<ActivityResult>
    fun parse(data: Any)
}

abstract class SimpleActivity : Activity {
    override fun proceed(ctx: ActivityContext): Observable<ActivityResult> {
        return doProceed(ctx).filter {
            return@filter (it is ActivityResult)
        }.map {
            it as ActivityResult
        }.toObservable()
    }

    override fun parse(data: Any) {
        doParse(data)
    }

    abstract fun doProceed(ctx: ActivityContext): Maybe<out Any>
    abstract fun doParse(data: Any)
}

class EchoActivity(val message: String) : SimpleActivity() {
    override fun doParse(data: Any) {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    companion object {
        val logger = LogFactory.getLog(EchoActivity::class.java)
    }

    override fun doProceed(ctx: ActivityContext): Maybe<out Any> {
        return Maybe.empty<Any>().doOnComplete {
            logger.info(message)
        }
    }
}

class ConnectActivity(private val grpc: GRPC) : SimpleActivity() {
    override fun doParse(data: Any) {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    val address = ""
    val port = 0
    override fun doProceed(ctx: ActivityContext): Maybe<out Any> {
        if (ctx !is RobotActivityContext) {
            return Maybe.error(RuntimeException())
        }

        val p = Common.ConnectParams.newBuilder()
                .setAddress(address)
                .setConnectionId(Common.ConnectionIdentity.newBuilder().setAccount(ctx.robot.account))
                .build()
        return grpc.connect(p).toMaybe()
    }

}

open class CompositeActivity : Activity {
    override fun parse(data: Any) {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun proceed(ctx: ActivityContext): Observable<ActivityResult> {
        return Observable.fromIterable(activities).concatMap {
            it.proceed(ctx)
        }
    }

    val activities: List<Activity> = mutableListOf()


}

class LoopActivity : CompositeActivity() {
    private val loopTimes = 1L

    override fun proceed(ctx: ActivityContext): Observable<ActivityResult> {
        return super.proceed(ctx).repeat(loopTimes)
    }
}

interface ActivityFactory {
    fun create(data: Any): Activity
}

@Component
class DefaultActivityFactory : ActivityFactory {
    override fun create(data: Any): Activity {
        val a = CompositeActivity()
        a.parse(data)
        return a
    }

}